modules/auxiliary/gather/ldap_query.rb

```diff
@@ -206,6 +206,21 @@ def output_data_csv(entries)
     generate_rex_tables(entries, 'csv')
   end
 
+  def normalize_entries(entries)
+    cleaned_entries = []
+    for entry in entries
+      # Convert to a hash so we get only the data we need.
+      entry = entry.to_h
+      entry_keys = entry.keys
+      for key in entry_keys
+        entry[key][0] = entry[key][0].force_encoding('ISO-8859-1').encode('UTF-8')
```

Comment:
Does the ldap server return ISO-8859-1 encoded strings? Is that something we've configured/negotiated with the ldap server? 👀 

From a super quick glance - it looks like the encoding may depend on the ldap version and whatever the target has been configured with - but let me know if I'm missing something here:

> When processing an LDAP v2 request, Active Directory exhibits the following behavioral differences from processing an LDAP v3 request:
>
> - Instead of using the [UTF-8](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/b645c125-a7da-4097-84a1-2fa7cea07714#gt_409411c4-b4ed-4ab6-b0ee-6d7815f85a35) character encoding for LDAPString [[RFC2251]](https://go.microsoft.com/fwlink/?LinkId=90325), the system's configured [code page](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/b645c125-a7da-4097-84a1-2fa7cea07714#gt_210637d9-9634-4652-a935-ded3cd434f38) is used. The code page is configured locally on the DC by the DC's administrator.
> https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/5b4872af-d421-42ff-9e47-2ce17159a965

I'm also wondering if we're missing other data types here - such as timestamps etc, or is that handled transparently by the ldap library already?

Reply:
I was originally envisioning a solution involving [`Rex::Text#to_hex_ascii`](https://github.com/rapid7/rex-text/blob/master/lib/rex/text/hex.rb#L41) which would print the "original string with non-printables converted to their escaped hex representation".

Reply:
I think Spencer's solution here would likely be the best way forwards vs assuming an encoding set. I wasn't 100% convinced this solution was the best at the time but was posting an initial working implementation to get some feedback on this. As for the timestamp comment etc, see Spencer's comment below. Will look to address those in future PRs.

Reply:
Output with Spencer's suggestion incorporated:
```
msf6 auxiliary(gather/ldap_query) > run
[*] Running module against 172.20.10.51

[+] Successfully bound to the LDAP server!
[*] Discovering base DN automatically
[+] 172.20.10.51:389 Discovered base DN: DC=daforest,DC=com
[*] CN=daforest-WIN-BR0CCBA815B-CA CN=Enrollment Services CN=Public Key Services CN=Services CN=Configuration DC=daforest DC=com
{
  "dn": "CN=daforest-WIN-BR0CCBA815B-CA,CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration,DC=daforest,DC=com",
  "cn": "daforest-WIN-BR0CCBA815B-CA",
  "cacertificate": "0\\x82\\x03\\x850\\x82\\x02m\\xa0\\x03\\x02\\x01\\x02\\x02\\x10-v\\xba\\x03H\\x06\\xea\\xb1G\\xb0\\xa0\\x93{\\xb7\\xea\\x1a0\\x0d\\x06\\x09*\\x86H\\x86\\xf7\\x0d\\x01\\x01\\x0b\\x05\\x000U1\\x130\\x11\\x06\\x0a\\x09\\x92&\\x89\\x93\\xf2,d\\x01\\x19\\x16\\x03com1\\x180\\x16\\x06\\x0a\\x09\\x92&\\x89\\x93\\xf2,d\\x01\\x19\\x16\\x08daforest1$0\"\\x06\\x03U\\x04\\x03\\x13\\x1bdaforest-WIN-BR0CCBA815B-CA0\\x1e\\x17\\x0d220825131802Z\\x17\\x0d270825132802Z0U1\\x130\\x11\\x06\\x0a\\x09\\x92&\\x89\\x93\\xf2,d\\x01\\x19\\x16\\x03com1\\x180\\x16\\x06\\x0a\\x09\\x92&\\x89\\x93\\xf2,d\\x01\\x19\\x16\\x08daforest1$0\"\\x06\\x03U\\x04\\x03\\x13\\x1bdaforest-WIN-BR0CCBA815B-CA0\\x82\\x01\"0\\x0d\\x06\\x09*\\x86H\\x86\\xf7\\x0d\\x01\\x01\\x01\\x05\\x00\\x03\\x82\\x01\\x0f\\x000\\x82\\x01\\x0a\\x02\\x82\\x01\\x01\\x00\\x9e\\xe2\\x95\\xd5K\\x99\\x04\\x8b0\\xdf\\xb27\\x124\\xcb6\\xad\\xa7\\x9cy\\x03^i\\x8f\\x04\\xd5\\xb58XH\\xfc\\x1f:z\\x02\\x06\\x0d\\x110\\xe0\\xa0\\xd3\\xdd\\x83\\xf4\\xdcn(\\xc8m;N\\xf1F\\xf6\\xb8\\xb5#\\xa5\\x82\\x97v&\\xe6\\x1438v\\xc0\\xe3\\xd3wSP\\x0c\\xf1\\x95\\xeb\\x9f\\xd2\\xb5\\xb2\\x908\\x81~2\\x08\\x0e\\xf9\\xd6<<\\xe8\\xbdH,\\x10\\x98/\\x9e\\x19\\xf6\\xe7&\\xd3\\x95x^\\xa1\\xbc64.\\xfc-h7\\xab \\xf8\\x1d\\x18\\x0fG\\x0e\\x86\\xde8\\xabD\\xcaNV\\xda\\xaa\\x8e}\\xe8t0R5\\xe0\\xa1K\\xba3\\x1fKm\\x9b\\x05\\xc4\\xed\\x88\\x9c\\x80U-\\xb7\\x0e\\xfd\\x19\\xdf\\x88\\xe5\\x1c6\\x14OF1\\x05\\xcc`\\xb6\\xe8\\xdc\\xd2\\xb2\\x85`\\xc4\\x93\\xf6K\\x11fc\\xdd\\xb4\\xc6\\xd3P\\xca\\xbc*>hG \\xb4\\x94\\x1d\\xef\\xf0\\xb32\\xe5&S\\xae\\x9b\\xcc\\x09\\xcb\\xef\\xfd\\xff\\xed\\x86\\x8e\\xa0E\\xe5p=\\xf9?\\xbe\\n\\xe3\\x0cj\\xbf\\x02\\xfb\\xb7:\\xad\\xef\\xb52D\\xc7\\xf8\\x83\\x92(\\x03\\xaaDj\\x85\\x02\\x03\\x01\\x00\\x01\\xa3Q0O0\\x0b\\x06\\x03U\\x1d\\x0f\\x04\\x04\\x03\\x02\\x01\\x860\\x0f\\x06\\x03U\\x1d\\x13\\x01\\x01\\xff\\x04\\x050\\x03\\x01\\x01\\xff0\\x1d\\x06\\x03U\\x1d\\x0e\\x04\\x16\\x04\\x14\\xdd\\x1c\\x7f\\xd8\\xee\\xf9\\xa8\\xd0\\xaan\\xacUa\\xe0p\\xacE\\x03pl0\\x10\\x06\\x09+\\x06\\x01\\x04\\x01\\x827\\x15\\x01\\x04\\x03\\x02\\x01\\x000\\x0d\\x06\\x09*\\x86H\\x86\\xf7\\x0d\\x01\\x01\\x0b\\x05\\x00\\x03\\x82\\x01\\x01\\x00T\\xc6\\x16\\xd7\\x85\\xd1Np\\xac\\xaa\\x17R\\x92\\xc3\\xaa\\xc6\\xf7\\xae\\x07\\xc9}y\\Q,\\xdb\\xa3\\x99q\\x01fv\\xd7\\xc6%\\xc9\\x95\\x11\\x10\\xadtk{\\x05\\xf22\\x07v\\x97]7?\\x13\\x90\\xd7tD\\x98\"<\\xa9\\xf3Z\\xf9\\x13\\x19^J\\xe8\\x9f\\x80(\\xa5G\\x0d\\xedU\\x16lC\\x00\\xde\\x8c\\x97*\\x86'\\x06\\xde?i\\x1e\\x1aU\\xf0\\xc1\\xe7\\xa4\\x04\\xc0\\x8a\\xc2\\xac\\xa5\\xc8b\\xc9\\x09\\xba\\xf1\\xa79\\x0e\\x0a\\x84\\x1al\\x8dj\\x1f4\\x1f\\xb5\\xbd\\xe5\\xc2\\x8fm\\x7fw\\xcf\\xe0\\xc6:C%\\xd0\\xc9\\x83\\x89\\x1d\\xf6\\xfa\\xca\\xeb\\xd7\\x96&(\\xea\\x8d\\x19\\x1b\\xc3\\xd7J\\x05\\x8b]\\xd3\\x0cj\\x8ep\\x1c\\xe75i_\\xcf\\xcfmQ\\xed\\xb85,\\xb4\\xb4x\\x06\\x90V\\xb1\\xf1o\\xe8\\x8b\\xab\\x08\\xfa\\x8f\\x84\\xa4\\xd2\\xa9\\x99K,\\xc8\\x01K\\xf4\\x1d\\x9fd\\xc0?qAL\\x17\\xbb\\x11O\\xa1\\x88\\x9cw\\xb9\\xe1\\xae\\xfe1\\xb1\\x84\\x87\\x12\\x85\\xab\\x12\\xfc\\xd1\\xf0\\x96\\x1f\\xb9<')\\x98\\xd5G\\x10\\xf6(\\x05o\\x85\\x0e\\xb36D\\xc3\\xea.",
  "name": "daforest-WIN-BR0CCBA815B-CA",
  "objectguid": "\"\\x8f6%HL\\xc2D\\xbd\\x07c\\xefxl\\xbf\\xaa",
  "dnshostname": "WIN-BR0CCBA815B.daforest.com",
  "cacertificatedn": "CN=daforest-WIN-BR0CCBA815B-CA, DC=daforest, DC=com",
  "certificatetemplates": "DirectoryEmailReplication || DomainControllerAuthentication || KerberosAuthentication || EFSRecovery || EFS || DomainController || WebServer || Machine || User || SubCA || Administrator"
}
[*] Auxiliary module execution completed
msf6 auxiliary(gather/ldap_query) > 
```